src/base/lib/Convex/BuildTx.hs

```diff
@@ -108,71 +113,78 @@ import           Convex.MonadLog            (MonadLog (..), MonadLogIgnoreT,
+import           Data.Maybe                 (fromMaybe, fromJust)
 import qualified Data.Set                   as Set
 import qualified Plutus.V1.Ledger.Api       as Plutus
 
-{-| A map of all inputs of the final transaction produced by a @TxBuilder@
--}
--- TOOO: This is essentially a subset of @TxBodyContent BuildTx@. Maybe we should
---       just make the entire @TxBodyContent BuildTx@ available to 'addInput'?
-data TransactionInputs = TransactionInputs
-  { txiSpendingInputs  :: Map C.TxIn (Witness WitCtxTxIn C.BabbageEra) -- ^ Inputs spent by the final transaction
-  , txiReferenceInputs :: Set C.TxIn -- ^ Reference inputs used by the final transaction
-  }
-
-mkTxInputs :: C.TxBodyContent C.BuildTx C.BabbageEra -> TransactionInputs
-mkTxInputs TxBodyContent{txIns, txInsReference} =
-  TransactionInputs
-    { txiSpendingInputs  = Map.fromList (fmap (view L._BuildTxWith) <$> txIns)
-    , txiReferenceInputs = Set.fromList (view L._TxInsReference txInsReference)
-    }
+type TxBody = C.TxBodyContent C.BuildTx C.BabbageEra
 
 {-| Look up the index of the @TxIn@ in the list of spending inputs
 -}
-lookupIndexSpending :: C.TxIn -> TransactionInputs -> Maybe Int
-lookupIndexSpending txi = Map.lookupIndex txi . txiSpendingInputs
+lookupIndexSpending :: C.TxIn -> TxBody -> Maybe Int
+lookupIndexSpending txi = Map.lookupIndex txi . Map.fromList . (fmap (view L._BuildTxWith) <$>) . view L.txIns
+
+{-| Look up the index of the @TxIn@ in the list of spending inputs. Throws an error if the @TxIn@ is not present.
+-}
+findIndexSpending :: C.TxIn -> TxBody -> Int
+findIndexSpending txi = fromJust . lookupIndexSpending txi
 
 {-| Look up the index of the @TxIn@ in the list of reference inputs
 -}
-lookupIndexReference :: C.TxIn -> TransactionInputs -> Maybe Int
-lookupIndexReference txi = Set.lookupIndex txi . txiReferenceInputs
+lookupIndexReference :: C.TxIn -> TxBody -> Maybe Int
+lookupIndexReference txi = Set.lookupIndex txi . Set.fromList . view (L.txInsReference . L._TxInsReference)
 
-{-| Look up the index of the @TxIn@ in the list of spending inputs
+{-| Look up the index of the @TxIn@ in the list of reference inputs. Throws an error if the @TxIn@ is not present.
 -}
-findIndexSpending :: C.TxIn -> TransactionInputs -> Int
-findIndexSpending txi = Map.findIndex txi . txiSpendingInputs
+findIndexReference :: C.TxIn -> TxBody -> Int
+findIndexReference txi = fromJust . lookupIndexReference txi
 
-{-| Look up the index of the @TxIn@ in the list of reference inputs
+{-| Look up the index of the @PolicyId@ in the transaction mint.
+Note: cardano-api represents a value as a @Map AssetId Quantity@, this is different than the on-chain representation
+which is @Map CurrencySymbol (Map TokenName Quantity).
+Here, we want to get the index into the on-chain map, but instead index into the cardano-api @Map CurrencySymbol Witness@.
+These two indexes should be the same by construction, but it is possible to violate this invariant when building a tx.
```

Comment:
I think it might be possible to violate the invariant when building the Tx, but I don't think we could ever have a situation where such a Tx was able to actually execute on chain. I'm not positive, though; maybe we should try it in an integration test.

Reply:
Agree that only invalid transactions can violate the invariant, but what I'm wondering is: Can this construction prevent us from building a valid transaction in a case where it should be possible to build one?